### PR TITLE
Extend taint tracking interface with flow states

### PR DESCRIPTION
**Pull request by jketema**
The flow state extension had to not been propagated to taint analysis, which lead to (a) a weird documentation mix and (b) having to use `isBarrier` and `isAdditionalFlowStep` in places where the user would expect `isSanitizer` and `isAdditionalTaintStep`. This PR fixes this.

I had to fix two Python libraries, a Java library and a Java query, which makes me somewhat worried this change comes a bit too late. Some advice on that would be useful. One solution would be to not mark `isBarrier` and `isAdditionalFlowStep` as final.

### CODE REVIEW

**Comment by atorralba** [comment]
The Java changes LGTM 👍

As the author of the Java query using flow states, this is what I would've liked to find in the taint tracking library when I used it 😄

**Review by yoff — DISMISSED** [review]
Lgtm

**Review by hvitved — DISMISSED** [review]
LGTM

File: cpp/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md [hunk 1]
```diff
@@ -0,0 +1,4 @@
+---
+category: breaking
+---
+* The flow state variants of `isBarrier` and `isAdditionalFlowStep` are no longer exposed in the taint tracking library. The `inSanitizer` and `isAdditionalTaintStep` predicates should be used instead.
```
Comment (atorralba) [inline cpp/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md:4]: Typo in all the change notes:

```suggestion
* The flow state variants of `isBarrier` and `isAdditionalFlowStep` are no longer exposed in the taint tracking library. The `isSanitizer` and `isAdditionalTaintStep` predicates should be used instead.
```

Comment (jketema) [reply cpp/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md:4]: I'm, not awake yet apparently. Thanks for spotting that.

Comment (jketema) [reply cpp/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md]: Fixed.

File: python/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md [hunk 2]
```diff
@@ -0,0 +1,4 @@
+---
+category: breaking
+---
+* The flow state variants of `isBarrier` and `isAdditionalFlowStep` are no longer exposed in the taint tracking library. The `inSanitizer` and `isAdditionalTaintStep` predicates should be used instead.
```
Comment (atorralba) [inline python/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md:4]:
```suggestion
* The flow state variants of `isBarrier` and `isAdditionalFlowStep` are no longer exposed in the taint tracking library. The `isSanitizer` and `isAdditionalTaintStep` predicates should be used instead.
```

Comment (jketema) [reply python/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md:4]: 🤦‍♂️

Comment (jketema) [reply python/ql/lib/change-notes/2022-03-14-taint-interface-cleanup.md]: Three times is a charm?

**Review by atorralba — APPROVED** [review]

**Review by MathiasVP — APPROVED** [review]

**Pull request merged**